### PR TITLE
One formatter for Timestamp fields in Storage writes

### PR DESCRIPTION
**Pull request by ahmedabu98**
Allow timestamp fields with UTC suffix

e.g. "2023-02-14 12:00:00.123 UTC"

Also, change to use one date time formatter

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java [hunk 1]
```diff
@@ -746,6 +755,12 @@ private static void fieldDescriptorFromTableField(
               return ChronoUnit.MICROS.between(
                   Instant.EPOCH, Instant.ofEpochMilli(Long.parseLong((String) value)));
             } catch (NumberFormatException e2) {
+              if (((String) value).endsWith("UTC")) {
```
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:758]: we should not catch an exception to determine we should use an alternate date time formatter

Comment (ahmedabu98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:758]: that was already the current implementation, will change to have one date time formatter to parse from.

**Comment by ahmedabu98** [comment]
R: @johnjcasey PTAL

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java [hunk 2]
```diff
@@ -82,6 +82,17 @@ public class TableRowToStorageApiProto {
           .append(DateTimeFormatter.ISO_LOCAL_DATE)
           .appendLiteral(' ')
           .append(DateTimeFormatter.ISO_LOCAL_TIME)
+          .optionalStart()
+          .appendLiteral(" UTC")
```
Comment (an2x) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:86]: I think we should support any time zone here, not just UTC.
Even better, support all formats listed in [BigQuery documentation](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#canonical_format_for_timestamp_literals), i.e.:
* 2014-09-27T12:30:00.45Z
* 2014-09-27 12:30:00.45-8:00
* 2014-09-27 12:30:00.123456 UTC
* 2014-09-27 12:30:00.123456 America/Los_Angeles
Note that in all 4 formats BQ accepts either space or "T" as the date/time separator. And as for the datetime/time zone separator: for "Z" and "-8:00" time zone format there must be no space before the time zone, and for "UTC" and "America/Los_Angeles" format there must be a space before.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java [hunk 3]
```diff
@@ -80,11 +81,32 @@ public class TableRowToStorageApiProto {
   private static final DateTimeFormatter DATETIME_SPACE_FORMATTER =
       new DateTimeFormatterBuilder()
           .append(DateTimeFormatter.ISO_LOCAL_DATE)
+          .optionalStart()
           .appendLiteral(' ')
+          .optionalEnd()
+          .optionalStart()
+          .appendLiteral('T')
+          .optionalEnd()
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:89]: FYI this allows for timestamp fields in TableRows to look like this:
1. `2000-02-13T12:30:00`
2. `2000-02-13 12:30:00`
3. `2000-02-13 T12:30:00`
4. `2000-02-1312:30:00`

All of these would be handled the same way so I don't think there is much concern. If there's an issue however, I can split up the DATETIME_SPACE_FORMATTER so only 1) and 2) are valid

**Comment by ahmedabu98** [comment]
@johnjcasey, @an2x read for review

**Comment by johnjcasey** [comment]
LGTM

**Comment by apilloud** [comment]
R: @johnjcasey @an2x

**Pull request merged**